spring-xd-reactor/src/main/java/org/springframework/xd/reactor/MultipleBroadcasterMessageHandler.java

```diff
@@ -0,0 +1,196 @@
+ * dispatcher thread will be the same for each partition.
+ * <p/>
+ * If the Stream mapped to the partitionExpression value has an error or completes, it will be recreated when the
+ * next message consumed maps to the same partitionExpression value.
+ * <p/>
+ * All error handling is the responsibility of the processor implementation.
+ *
+ * @author Mark Pollack
+ */
+public class MultipleBroadcasterMessageHandler extends AbstractMessageProducingHandler implements DisposableBean,
+        IntegrationEvaluationContextAware {
+
+    protected final Log logger = LogFactory.getLog(getClass());
+
+    private final ConcurrentMap<Object, Broadcaster<Object>> broadcasterMap =
+            new ConcurrentHashMap<Object, Broadcaster<Object>>();
+
+    private final Map<Object, Controls> controlsMap = new Hashtable<Object, Controls>();
+
+    private final Environment environment;
+
+    @SuppressWarnings("rawtypes")
+    private final Processor processor;
+
+    private final ResolvableType inputType;
+
+    private final Expression partitionExpression;
+
+    private final SpelExpressionParser spelExpressionParser = new SpelExpressionParser();
+
+    private EvaluationContext evaluationContext = new StandardEvaluationContext();
+
+    /**
+     * Construct a new SynchronousDispatcherMessageHandler given the reactor based Processor to delegate
+     * processing to.
+     *
+     * @param processor The stream based reactor processor
+     */
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    public MultipleBroadcasterMessageHandler(Processor processor, String partitionExpression) {
+        Assert.notNull(processor, "processor cannot be null.");
+        Assert.notNull(partitionExpression, "Partition expression can not be null");
+        this.processor = processor;
+        this.partitionExpression = spelExpressionParser.parseExpression(partitionExpression);
+        environment = Environment.initializeIfEmpty(); // This by default uses SynchronousDispatcher
+        Method method = ReflectionUtils.findMethod(this.processor.getClass(), "process", Stream.class);
+        this.inputType = ResolvableType.forMethodParameter(method, 0).getNested(2);
+    }
+
+    @Override
+    protected void handleMessageInternal(Message<?> message) {
+        Broadcaster<Object> broadcasterToUse = getBroadcaster(message);
+        if (ClassUtils.isAssignable(inputType.getRawClass(), message.getClass())) {
+            broadcasterToUse.onNext(message);
+        } else if (ClassUtils.isAssignable(inputType.getRawClass(), message.getPayload().getClass())) {
+            broadcasterToUse.onNext(message.getPayload());
+        } else {
+            throw new MessageHandlingException(message, "Processor signature does not match [" + message.getClass()
+                    + "] or [" + message.getPayload().getClass() + "]");
+        }
+
+        if (logger.isDebugEnabled()) {
+            Controls controls = this.controlsMap.get(Thread.currentThread().getId());
```

Comment:
What I'm saying is that this map is populated by using the SpEL partition expression. There is no reason that it maps to either a String, let alone a Thread id